ffi/src/transaction/mod.rs

```diff
@@ -0,0 +1,355 @@
+//! This module holds functionality for managing transactions.
+mod write_context;
+
+use crate::error::{ExternResult, IntoExternResult};
+use crate::handle::Handle;
+use crate::unwrap_and_parse_path_as_url;
+use crate::KernelStringSlice;
+use crate::{DeltaResult, ExternEngine, Table, Url};
+use crate::{ExclusiveEngineData, SharedExternEngine};
+use delta_kernel::transaction::{CommitResult, Transaction};
+use delta_kernel_ffi_macros::handle_descriptor;
+
+/// A handle representing an exclusive transaction on a Delta table.
+///
+/// This struct provides a safe wrapper around the underlying `Transaction` type,
+/// ensuring exclusive access to transaction operations. The transaction can be used
+/// to stage changes and commit them atomically to the Delta table.
+#[handle_descriptor(target=Transaction, mutable=true, sized=true)]
+pub struct ExclusiveTransaction;
+
+/// Start a transaction on the latest snapshot of the table.
+///
+/// # Safety
+///
+/// Caller is responsible for passing valid handles and path pointer.
+#[no_mangle]
+pub unsafe extern "C" fn transaction(
+    path: KernelStringSlice,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<ExclusiveTransaction>> {
+    let url = unsafe { unwrap_and_parse_path_as_url(path) };
+    let engine = unsafe { engine.as_ref() };
+    transaction_impl(url, engine).into_extern_result(&engine)
+}
+
+fn transaction_impl(
+    url: DeltaResult<Url>,
+    extern_engine: &dyn ExternEngine,
+) -> DeltaResult<Handle<ExclusiveTransaction>> {
+    let table = Table::try_from_uri(url?)?;
+    let transaction = table.new_transaction(extern_engine.engine().as_ref())?;
+    Ok(Box::new(transaction).into())
+}
+
+/// # Safety
+///
+/// Caller is responsible for passing a valid handle.
+#[no_mangle]
+pub unsafe extern "C" fn free_transaction(txn: Handle<ExclusiveTransaction>) {
+    txn.drop_handle();
+}
+
+/// Attaches commit information to a transaction. The commit info contains metadata about the
+/// transaction that will be written to the log during commit.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid handle. CONSUMES TRANSACTION and commit info
+#[no_mangle]
+pub unsafe extern "C" fn with_commit_info(
+    txn: Handle<ExclusiveTransaction>,
```

Comment:
ah - also noticing that the builder pattern is a bit odd over FFI? probably don't need to immediately solve it (unless anyone has other ideas right now) but could make an issue to see if we can make this feel nicer? @samansmink WDYT of this overall API structure? Is it amenable to C++ or annoying?